debugger/src/peripherals/svd_variables.rs

```diff
@@ -1,18 +1,303 @@
-use std::fmt::Debug;
-
 use crate::DebuggerError;
 use probe_rs::debug::VariableCache;
-use svd_rs::{peripheral, Device, MaybeArray::Single, PeripheralInfo};
+use std::{any, fmt::Debug, fs::File, io::Read, path::PathBuf};
+use svd_parser::{self as svd, ValidateLevel};
+use svd_rs::{Device, EnumeratedValues, FieldInfo, PeripheralInfo, RegisterInfo};
 
 /// The SVD file contents and related data
 #[derive(Debug)]
 pub(crate) struct SvdCache {
     /// A unique identifier
     pub(crate) id: i64,
-    /// The Device file represents the top element in a SVD file.
-    pub(crate) svd_device: Device,
     /// The SVD contents and structure will be stored as variables, down to the Register level.
     /// Unlike other VariableCache instances, it will only be built once per DebugSession.
     /// After that, only the SVD fields change values, and the data for these will be re-read everytime they are queried by the debugger.
     pub(crate) svd_registers: VariableCache,
 }
+
+impl SvdCache {
+    /// Create the SVD cache for a specific core. This function loads the file, parses it, and then builds the VariableCache.
+    pub(crate) fn new(svd_file: &PathBuf) -> Result<Self, DebuggerError> {
+        let svd_xml = &mut String::new();
+        match File::open(svd_file.as_path()) {
+            Ok(mut svd_opened_file) => {
+                svd_opened_file.read_to_string(svd_xml);
+                match svd::parse(&svd_xml) {
+                    Ok(peripheral_device) => Ok(SvdCache {
+                        id: probe_rs::debug::get_sequential_key(),
+                        svd_registers: variable_cache_from_svd(peripheral_device),
+                    }),
+                    Err(error) => Err(DebuggerError::Other(anyhow::anyhow!(
+                        "Unable to parse CMSIS-SVD file: {:?}. {:?}",
+                        svd_file,
+                        error,
+                    ))),
+                }
+            }
+            Err(error) => Err(DebuggerError::Other(anyhow::anyhow!("{}", error))),
+        }
+    }
+}
+
+/// Create a [`probe_rs::debug::VariableCache`] from a Device that was parsed from a CMSIS-SVD file.
+pub(crate) fn variable_cache_from_svd(peripheral_device: Device) -> probe_rs::debug::VariableCache {
+    probe_rs::debug::VariableCache::new()
+}
+
+/// Resolve all the peripherals through their (optional) `derived_from` peripheral.
+pub(crate) fn resolve_peripherals(
+    peripheral_device: &Device,
+) -> Result<Vec<PeripheralInfo>, DebuggerError> {
+    let mut resolved_peripherals = vec![];
+    for device_peripheral in &peripheral_device.peripherals {
+        // TODO: Need to code for the impact of MaybeArray results.
+        let mut peripheral_builder = PeripheralInfo::builder();
+        if let Some(derived_from) = &device_peripheral.derived_from {
+            if let Some(template_peripheral) = peripheral_device.get_peripheral(derived_from) {
```

Comment:
You could activate "derive-from" feature for `svd-rs`

Reply:
@burrbull ... Thank you, that sounds like an interesting and useful feature. Now that you mentioned it, I can see it in that crate's Cargo.toml, but can't seem to find any docs for it. Can you point me to it please? Or I can just experiment LOL

Reply:
Possibly it's because I often forget to add `--all-features` when publish. Maybe it is time to add `derived-from` to `defaults`.

Reply:
Usage in `svd2rust`:
https://github.com/rust-embedded/svd2rust/blob/a5f8239b77e81dbdef28171fbaa8268b44a7350b/src/generate/peripheral.rs#L35

Reply:
Nice and clean!  Thank you :)

Reply:
Implemented. Thanks again for your help!